src/geometry/FASER2DetectorConstruction.cc

```diff
@@ -21,152 +21,239 @@ FASER2DetectorConstruction::FASER2DetectorConstruction()
 
-  } else if ( opt == GeometricalParameters::magnetOption::CrystalPulling ){
- 
-    G4cout << "Building CrystalPulling spectrometer magnet" << G4endl;
-    fMagnetLengthZ = GeometricalParameters::Get()->GetFASER2MagnetLengthZ();
-    fMagnetInnerR = GeometricalParameters::Get()->GetFASER2MagnetInnerR();
-    fMagnetOuterR = GeometricalParameters::Get()->GetFASER2MagnetOuterR();
-    fNMagnets = GeometricalParameters::Get()->GetNFASER2Magnets();
-    fMagnetGap = GeometricalParameters::Get()->GetFASER2MagnetGap();
+  fEMCaloThickness = GeometricalParameters::Get()->GetFASER2EMCaloThickness();
+  fHadCaloThickness = GeometricalParameters::Get()->GetFASER2HadCaloThickness();
+  fIronWallThickness = GeometricalParameters::Get()->GetFASER2IronWallThickness();
+
+  fVetoLengthX = GeometricalParameters::Get()->GetFASER2VetoLengthX();
+  fVetoLengthY = GeometricalParameters::Get()->GetFASER2VetoLengthY();
+  fVetoShieldThickness = GeometricalParameters::Get()->GetFASER2VetoShieldThickness();
+
+  //* Work out the maximum extent of the whole of the FASER2 detector so we can make the physical volume
+  //* We want to avoid any overlaps as this can be problematic for GENIE when it reads the GDML file
+  G4double vetoLength = 2*fScinThickness + fVetoShieldThickness;
+  G4double upstreamTrackerLength = fNUpstreamTrackers*fScinThickness + (fNUpstreamTrackers - 1)*fTrackingStationGap + fUpstreamTrackingStationGap;
+  G4double magnetLength = fNMagnets*fMagnetLengthZ + (fNMagnets-1)*fMagnetGap;
+  G4double downstreamTrackerLength = fNDownstreamTrackers*fScinThickness + (fNDownstreamTrackers - 1)*fTrackingStationGap + fDownstreamTrackingStationGap;
+  G4double caloLength = fTrackingStationGap+fEMCaloThickness+fHadCaloThickness+fIronWallThickness;
+  G4double muonLength = fScinThickness+fTrackingStationGap+0.5*m;
+  
+  G4double totLengthZ =  vetoLength + fDecayVolumeLength +  upstreamTrackerLength + magnetLength + downstreamTrackerLength + caloLength + muonLength;
+  
+  std::vector<G4double> lengths{fMagnetWindowX+2*fMagnetYokeThicknessX, fVetoLengthX, fMagnetOuterR};
+  std::vector<G4double> heights{fMagnetWindowY+2*fMagnetYokeThicknessY, fVetoLengthY, fMagnetOuterR};
+  
+  auto totLengthX = std::max_element(lengths.begin(), lengths.end());
+  auto totLengthY = std::max_element(heights.begin(), heights.end());
+
+  GeometricalParameters::Get()->SetFASER2TotalSizeZ(totLengthZ);
+  G4RotationMatrix *noRot = new G4RotationMatrix();
+  
+  //* Create the FASER2 volume - all components go in here
+  auto containerSolid = new G4Box("FASER2Solid", *totLengthX/2., *totLengthY/2., totLengthZ/2.);
+  fFASER2Assembly = new G4LogicalVolume(containerSolid, fMaterials->Material("Air"), "FASER2Logical");
+
+  G4VisAttributes* F2_logVisAtt = new G4VisAttributes(G4Colour::Grey());
+  F2_logVisAtt->SetForceWireframe(true);
+  F2_logVisAtt->SetForceSolid(false);
+  F2_logVisAtt->SetVisibility(true);
+  fFASER2Assembly->SetVisAttributes(F2_logVisAtt);
```

Comment:
What about moving all the visualization things at the end? Or in an external function that you call at the end? This is just to clean-up the code a bit. It seems that all of these logical volumes are class members, so it should be easy.